From faaddf0e0d3b5ad34eec575bd438ab9791d99295 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Thu, 8 Aug 2024 13:04:42 +0800
Subject: [PATCH 186/265] mtk: mt76: mt7996: add support for configuring
 multi-band CSA offset in per-STA profile

Add support for configuring multi-band CSA offset in per-STA profile.
This is a preliminary patch to support multi-band channel switch.
Please refer to https://gerrit.mediatek.inc/q/topic:%22multi_mlo_csa%22

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 mt7996/mcu.c | 6 ++++--
 mt7996/mcu.h | 6 +++---
 2 files changed, 7 insertions(+), 5 deletions(-)

diff --git a/mt7996/mcu.c b/mt7996/mcu.c
index 3b066a3e..ca88873e 100644
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -3834,6 +3834,7 @@ mt7996_mcu_beacon_sta_prof_csa(struct sk_buff *rskb,
 	struct mt7996_bss_conf *cs_mconf;
 	struct bss_bcn_sta_prof_cntdwn_tlv *sta_prof;
 	struct tlv *tlv;
+	u8 cs_band;
 
 	if (!ieee80211_vif_is_mld(vif) || !offs->sta_prof_cntdwn_offs[0])
 		return;
@@ -3844,9 +3845,10 @@ mt7996_mcu_beacon_sta_prof_csa(struct sk_buff *rskb,
 
 	tlv = mt7996_mcu_add_uni_tlv(rskb, UNI_BSS_INFO_BCN_STA_PROF_CSA, sizeof(*sta_prof));
 
+	cs_band = cs_mconf->phy->mt76->band_idx;
 	sta_prof = (struct bss_bcn_sta_prof_cntdwn_tlv *)tlv;
-	sta_prof->sta_prof_csa_offs = cpu_to_le16(offs->sta_prof_cntdwn_offs[0] - 4);
-	sta_prof->cs_bss_idx = cs_mconf->mt76.idx;
+	sta_prof->sta_prof_csa_offs[cs_band] = cpu_to_le16(offs->sta_prof_cntdwn_offs[0] - 4);
+	sta_prof->cs_bss_idx[cs_band] = cs_mconf->mt76.idx;
 }
 
 static void
diff --git a/mt7996/mcu.h b/mt7996/mcu.h
index 568dfd96..3ddab046 100644
--- a/mt7996/mcu.h
+++ b/mt7996/mcu.h
@@ -543,9 +543,9 @@ struct bss_bcn_crit_update_tlv {
 struct bss_bcn_sta_prof_cntdwn_tlv {
 	__le16 tag;
 	__le16 len;
-	__le16 sta_prof_csa_offs;
-	u8 cs_bss_idx;
-	u8 pkt_content[9];
+	__le16 sta_prof_csa_offs[__MT_MAX_BAND];
+	u8 cs_bss_idx[__MT_MAX_BAND];
+	u8 pkt_content[3];
 } __packed;
 
 struct bss_bcn_ml_reconf_tlv {
-- 
2.45.2

